Implement experimental @abi attribute

beccadax opened the pull request:
This is the beginning of an implementation of the `@abi` attribute [as recently pitched](https://forums.swift.org/t/pitch-controlling-the-abi-of-a-declaration/75123/8). It is intended to subsume the use of `@_silgen_name` in ABI compatibility hacks, providing a tool for handling API replacements and enhancements that is both more capable and usable by mere mortals.

~~This PR also replaces all standard library uses of `@_silgen_name` with a mangled Swift name with `@abi` uses. This is about a quarter of all standard library `@_silgen_name` uses (the remaining uses are to import or export runtime functions or compiler intrinsics by simple known names).~~ (Note: Standard library adoption will be deferred until the type checking rules are more stable.)

There's a lot left to do here, ~~but one of the most important things is making `@abi` on functions suppressible by providing a fallback to `@_silgen_name` in the module interface~~.

SwiftSyntax PR: swiftlang/swift-syntax#2882

beccadax committed:
[ClangImporter] Try to warn about missing types

beccadax committed:
[NFC] Dump an ASTScopeImpl’s parent nodes

beccadax committed:
Type check ABI decls

Sema now type-checks the alternate ABI-providing decls inside of @abi attributes.

Making this work—particularly, making redeclaration checking work—required making name lookup aware of ABI decls. Name lookup now evaluates both API-providing and ABI-providing declarations. In most cases, it will filter ABI-only decls out unless a specific flag is passed, in which case it will filter API-only decls out instead. Calls that simply retrieve a list of declarations, like `IterableDeclContext::getMembers()` and friends, typically only return API-providing decls; you have to access the ABI-providing ones through those.

As part of that work, I have also added some basic compiler interfaces for working with the API-providing and ABI-providing variants. `ABIRole` encodes whether a declaration provides only API, only ABI, or both, and `ABIRoleInfo` combines that with a pointer to the counterpart providing the other role (for a declaration that provides both, that’ll just be a pointer to `this`).

Decl checking of behavior specific to @abi will come in a future commit.

Note that this probably doesn’t properly exercise some of the new code (ASTScope::lookupEnclosingABIAttributeScope(), for instance); I expect that to happen only once we can rename types using an @abi attribute, since that will create distinguishable behavior differences when resolving TypeReprs in other @abi attributes.

beccadax committed:
Add some basic validation of vars and funcs

Check for:

• Matching decl kinds
• Matching PBD shapes (does every VarDecl on both sides have a counterpart?)
• Matching function effects
• Matching function arity (roughly)

beccadax committed:
Use @abi attribute in mangling

What’s implemented now is actually *far* more thorough than what the surface syntax can currently express, mainly because I can’t apply @abi to nominal types yet.

beccadax committed:
[NFC] Improve lookup flag printing facilities

beccadax committed:
Print @abi into module interface (suppressibly)

ktoso commented on stdlib/public/Distributed/DistributedActor.swift:
```diff
@@ -363,9 +363,13 @@ extension DistributedActor {
   // and even though this is @AEIC, the symbol name would conflict.
-  @_silgen_name("$s11Distributed0A5ActorPAAE20whenLocalTypedThrowsyqd__Sgqd__xYiYaYbqd_0_YKXEYaqd_0_YKs8SendableRd__s5ErrorRd_0_r0_lF")
+  @abi(
+    public nonisolated func __api_whenLocal<T: Sendable, E>(
```
Probably this would be nicer if it stayed using the explicit "explaining why this method is different" name like we did in the silgen name:
```suggestion
    public nonisolated func whenLocalTypedThrows<T: Sendable, E>(
```

beccadax replied:
There's actually three different variants. Here's what I think I'll do here:

```
  @abi(
    // We need to @abi here because the signature is the same as
    // `__separately_compiled_typed_throws_whenLocal(_:)`, and even though this
    // is @AEIC, the symbol name would conflict.
    public nonisolated func __typed_throws_whenLocal<T: Sendable, E>(
      _ body: @Sendable (isolated Self) async throws(E) -> T
    ) async throws(E) -> T?
  )
  @_alwaysEmitIntoClient
  public nonisolated func whenLocal<T: Sendable, E>(
    _ body: @Sendable (isolated Self) async throws(E) -> T
  ) async throws(E) -> T? { /* snip */ }

  // ABI: This is a workaround when in Swift 6 this method was introduced
  // in order to support typed-throws, but missed to add @_aeic.
  // In practice, this method should not ever be used by anyone, ever.
  @abi(
    public nonisolated func whenLocal<T: Sendable, E>(
      _ body: @Sendable (isolated Self) async throws(E) -> T
    ) async throws(E) -> T?
  )
  @usableFromInline
  nonisolated func __separately_compiled_typed_throws_whenLocal<T: Sendable, E>(
    _ body: @Sendable (isolated Self) async throws(E) -> T
  ) async throws(E) -> T? { /* snip */ }

  // ABI: Historical whenLocal, rethrows was changed to typed throws `throws(E)`
  @abi(
    public nonisolated func whenLocal<T: Sendable>(
      _ body: @Sendable (isolated Self) async throws -> T
    ) async rethrows -> T?
  )
  @usableFromInline
  nonisolated func __rethrows_whenLocal<T: Sendable>(
    _ body: @Sendable (isolated Self) async throws -> T
  ) async rethrows -> T? { /* snip */ }
```

ktoso replied:
This looks good, thank you @beccadax! I'm really enjoying this way to spell these

lorentey commented on stdlib/public/core/Optional.swift:
```diff
@@ -879,9 +884,15 @@ public func ?? <T: ~Copyable>(
 @_spi(SwiftStdlibLegacyABI) @available(swift, obsoleted: 1)
 @usableFromInline
-internal func ?? <T>(
+internal func _legacy_abi_optionalNilCoalescingOperator <T>(
```
Ooh, nice catch

lorentey reviewed (approved):
Thank you for working on this -- the core stdlib parts look good to me. (Great, in fact!)

beccadax committed:
[NFC-ish] Fix some crashes from early Decl::dump()

Changes dump output used in one test; otherwise NFC.

beccadax committed:
Diagnose use of @_silgen_name and @abi together

beccadax committed:
Tie attributes to language features

The new `DECL_ATTR_FEATURE_REQUIREMENT` macro in DeclAttr.def can be used to declare that an attribute should only be available when a related language feature is enabled.

Effects:

• `#if hasAttribute(someAttr)` will return `false` unless the required feature is enabled.
• Code completion will not include the attribute unless the required feature is enabled.
• `TypeChecker::checkDeclAttributes()` diagnoses non-implicit uses of the attribute.

Add this mechanism and use it to tie @abi to the ABIAttribute feature. Also design tests for it.

beccadax commented:
@swift-ci please test

beccadax commented:
I think the implementation here is getting towards something mergeable, if not fully usable, but I expect the feature to continue to evolve, including in some ways that will break source compatibility with the current version. (For example, I may ban access control keywords in an `@abi` attribute and instead inherit them from the attribute it's attached to.) For that reason, I'm planning to move the standard library adoption into its own PR before this comes out of draft and gets merged. The stdlib PR will continue to evolve as the compiler implementation improves.

beccadax commented:
The macOS test failure, at least, was caused by SwiftSyntax not being able to parse the new syntax, so I'm adding support for that.

beccadax commented:
With https://github.com/swiftlang/swift-syntax/pull/2882

@swift-ci please test

beccadax commented:
With https://github.com/swiftlang/swift-syntax/pull/2882

@swift-ci please test

bnbarham commented on test/IDE/complete_decl_attribute_feature_requirement.swift:
```diff
@@ -0,0 +1,154 @@
+// RUN: %target-swift-ide-test -code-completion -source-filename %s -code-completion-token=KEYWORD2 | %FileCheck %s -check-prefixes=KEYWORD2,KEYWORD2_DISABLED
```
Note that you can use `%batch-code-completion` here instead. It has the benefit of not needing the many RUN lines + can re-use the ast context. Thanks for adding these by the way, it's very much appreciated!

beccadax replied:
I had to modify the tool with a feature that could add suffixes to the check prefixes, but this definitely cleaned things up quite a bit!

bnbarham replied:
Oh nice, thank you! CC @rintaro

beccadax committed:
Parse and serialize `@abi` attribute

This attribute will allow you to specify an alternate version of the declaration used for mangling. It will allow minor adjustments to be made to declarations so long as they’re still compatible at the calling convention level, such as refining isolation or sendability, renaming without breaking ABI, etc.

The attribute is behind the experimental feature flag `ABIAttribute`.

beccadax commented:
@swift-ci please test

beccadax commented:
@swift-ci please test

rintaro commented on lib/ASTGen/Sources/ASTGen/DeclAttrs.swift:
```diff
@@ -17,12 +17,37 @@ import SwiftIfConfig
+        visitABIDecls(for: decl) { declToConnect in
+          abiAttr.connectToInverse(attachedTo: declToConnect)
+        }
+      }
```
Since this is a duplicated logic between C++ parser and ASTGen, is it possible to do this in AST? For example, make `Decl::setParsedAttrs(const DeclAttributes &)` and use it from C++ parser and ASTGen.

TBH, I'm not a fan of this API. This feels like `element.append(to: array)`

rintaro commented on lib/Parse/ParseDecl.cpp:
```diff
@@ -3427,6 +3427,67 @@ ParserStatus Parser::parseNewDeclAttribute(DeclAttributes &Attributes,
+      // to `connectToInverse()`.
+      if (isa<ValueDecl>(D)) {
+        auto inverseAttr = new (Context) ABIAttr(nullptr, /*isInverse=*/true,
+                                                 /*implicit=*/true);
```
We'd like to avoid this kind of implicit operation in Parser as much as possible. Is it possible to move this (and  `connectToInverse(Decl *)`) to a later stage? E.g. `TypeCheckAttr`.
That would also minimize the bridging interface for ASTGen.

rintaro commented on test/attr/attr_abi.swift:
Could you add some test cases with attributes and modifiers?
I'm particularly interested in the behavior for nested `@abi` attributes. i.e. `@abi(@abi(func foo()) ...`

beccadax replied:
In a future PR I will add very thorough logic for how `@abi` interacts with other attributes, including itself (`@abi` will be banned on an ABI-only declaration), so I'm going to leave this task aside for now.

beccadax committed:
[NFC] Extract autodiff parsing code to Decl.cpp

The `@differentiable` and `@derivative` attributes need a parent pointer. Move the code to populate it from Parser to AST so it can be more easily shared between the parsers.

Done in preparation for similar code to be added for `@abi`.

rintaro commented on lib/AST/Decl.cpp:
```diff
@@ -407,6 +408,19 @@ DeclAttributes Decl::getSemanticAttrs() const {
+  for (auto *attr : attrs.getAttributes<DerivativeAttr>())
+    attr->setOriginalDeclaration(this);
+  for (auto attr : attrs.getAttributes<ABIAttr, /*AllowInvalid=*/true>())
+    attr->createInverse(this);
```
Nice! Thank you for doing the autodiff things too!

beccadax replied:
Keep in mind that I haven't actually implemented these attributes in ASTGen, but you should be able to drop in an implementation for them now. :D

rintaro reviewed (commented):
Parser and ASTGen related changes LGTM!

beccadax committed:
[ASTGen] Lower `@abi` in ASTGen

beccadax committed:
[NFC] [ASTGen] Refactor attribute attachment

Use `Decl::attachParsedAttrs()` instead of `Decl::setAttrs()` to attach attributes to a declaration in ASTGen. This causes the common attribute-setup logic there to be run.

NFC in this commit because none of the attributes that have special setup logic are currently implemented in ASTGen. Prepares to add support for `@abi` in a future commit.

beccadax commented:
@swift-ci test

xymus commented on lib/Serialization/Deserialization.cpp:
```diff
@@ -6472,6 +6473,43 @@ llvm::Error DeclDeserializer::deserializeDeclCommon() {
+  if (ABIDeclCounterpartID != 0) {
+    // This decl is the `abiDecl` of an `ABIAttr`. Force the decl that `ABIAttr`
+    // belongs to so that we establish the relationship between them.
+    (void)MF.getDecl(ABIDeclCounterpartID);
```
Could we protect the two `MF.getDecl` here against errors by instead using `getDeclChecked`? Any error could either be bubbled up or dropped when safe to do so.

beccadax replied:
Might take a bit of work, but probably.

xymus replied:
Looks good, thank you!

beccadax commented on lib/Sema/TypeCheckAttr.cpp:
```diff
@@ -188,6 +188,143 @@ class AttributeChecker : public AttributeVisitor<AttributeChecker> {
+
+public:
+  void visitABIAttr(ABIAttr *attr) {
+    Decl *AD = attr->abiDecl;
```
Reviewers: In this PR, I'm mostly just trying to make sure I can find the right decls to check against one another. The exact rules I'm enforcing are extremely rough and incomplete placeholders; properly defining them will be the subject of my next PR on this feature.

xedin commented on include/swift/AST/ASTContext.h:
```diff
@@ -414,6 +415,19 @@ class ASTContext final {
+  /// `ABIAttr::abiDecl` to the declaration the `ABIAttr` is attached to.
+  ///
+  /// \seeAlso \c recordABIAttr()
+  llvm::DenseMap<Decl *, Decl *> ABIDeclCounterparts;
```
Have you considered turning this into a cached request instead?

beccadax replied:
This would be difficult to compute lazily because there's no way (other than this table) to walk up from an ABI-only declaration to its counterpart; you have to start from the counterpart and then walk down to the ABI-only decl. The most reliable way I've found to do that is to eagerly populate a table while parsing or deserializing. I suppose I could still create a dummy request whose `evaluate()` function returns the default output and then make `recordABIAttr()` manually cache non-default outputs, but I'm not sure if that would really buy us anything over what I'm doing here.

xedin replied:
I am suggesting it because we have a similar pattern with other requests that are cached manually after parse and deserialization, the primary advantage is use a common mechanism and storage.

beccadax replied:
Changed in 413c673da4128f8a10df200899647a609c91fec1.

xedin reviewed (approved):
Sema changes look good to me!

beccadax committed:
[NFC] Requestify ABI role computation

beccadax commented:
@swift-ci please test
